File: frontend/src/lib/van_sales.ts
import { apiClient } from "@/lib/api";

const BASE = "/api/v1/van-sales";

// ── Types ─────────────────────────────────────────────────────────────────────

export type VanStatus = "ACTIVE" | "INACTIVE" | "ON_ROUTE" | "IN_REPAIR";
export type TxnType = "SALE" | "RETURN" | "EXCHANGE" | "CREDIT";
export type TxnStatus = "DRAFT" | "CONFIRMED" | "INVOICED" | "CANCELLED" | "SYNCED" | "PENDING_SYNC";
export type PaymentStatus = "UNPAID" | "PARTIAL" | "PAID" | "REFUNDED";
export type PaymentMethod = "CASH" | "MPESA" | "CARD" | "CREDIT" | "MIXED";
export type VisitStatus = "PLANNED" | "CHECKED_IN" | "COMPLETED" | "MISSED" | "SKIPPED";
export type ReconciliationStatus = "OPEN" | "SUBMITTED" | "APPROVED" | "DISCREPANCY";

export interface Van {
  id: string;
  van_code: string;
  plate_number?: string;
  driver_id?: string;
  route_id?: string;
  warehouse_id?: string;
  capacity_kg?: number;
  status: VanStatus;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface VanStock {
  id: string;
  van_id: string;
  item_id: string;
  batch_id?: string;
  quantity: number;
  reserved_qty: number;
  unit_cost?: number;
  uom: string;
}

export interface VanVisit {
  id: string;
  van_id: string;
  route_id?: string;
  customer_id: string;
  route_date: string;
  sequence_no?: number;
  planned_time?: string;
  check_in_time?: string;
  check_out_time?: string;
  status: VisitStatus;
  gps_lat?: number;
  gps_lng?: number;
  missed_reason?: string;
  offline_id?: string;
  created_at: string;
}

export interface TxnLine {
  id: string;
  txn_id: string;
  item_id: string;
  batch_id?: string;
  uom: string;
  quantity: number;
  unit_price: number;
  discount_pct: number;
  discount_amt: number;
  tax_pct: number;
  tax_amt: number;
  line_total: number;
  return_reason?: string;
}

export interface VanPayment {
  id: string;
  txn_id: string;
  method: PaymentMethod;
  amount: number;
  reference_no?: string;
  collected_at: string;
  receipt_no?: string;
  notes?: string;
  created_at: string;
}

export interface VanSalesTxn {
  id: string;
  txn_no: string;
  van_id: string;
  route_id?: string;
  customer_id: string;
  visit_id?: string;
  driver_id?: string;
  txn_type: TxnType;
  txn_datetime: string;
  status: TxnStatus;
  payment_status: PaymentStatus;
  subtotal: number;
  discount_amount: number;
  tax_amount: number;
  total_amount: number;
  paid_amount: number;
  balance_due: number;
  currency: string;
  offline_id?: string;
  synced_at?: string;
  notes?: string;
  created_at: string;
  lines: TxnLine[];
  payments: VanPayment[];
}

export interface VanReconciliation {
  id: string;
  van_id: string;
  route_date: string;
  status: ReconciliationStatus;
  opening_stock?: Record<string, number>;
  closing_stock?: Record<string, number>;
  sales_qty?: Record<string, number>;
  returns_qty?: Record<string, number>;
  discrepancy?: Record<string, number>;
  total_sales?: number;
  total_cash?: number;
  total_mpesa?: number;
  total_credit?: number;
  total_returns?: number;
  submitted_by?: string;
  approved_by?: string;
  approved_at?: string;
  notes?: string;
  created_at: string;
}

export interface VSAIRec {
  id: string;
  agent_type: string;
  status: string;
  severity: string;
  title: string;
  body: string;
  data?: Record<string, unknown>;
  van_id?: string;
  driver_id?: string;
  route_id?: string;
  created_at: string;
}

export interface SyncResult {
  visits_created: number;
  txns_created: number;
  payments_created: number;
  errors: string[];
  duplicates_skipped: number;
}

// ── Color helpers ─────────────────────────────────────────────────────────────

export const VAN_STATUS_COLORS: Record<VanStatus, string> = {
  ACTIVE:    "bg-emerald-500/20 text-emerald-300",
  INACTIVE:  "bg-slate-500/20 text-slate-400",
  ON_ROUTE:  "bg-blue-500/20 text-blue-300",
  IN_REPAIR: "bg-amber-500/20 text-amber-300",
};

export const VISIT_STATUS_COLORS: Record<VisitStatus, string> = {
  PLANNED:    "bg-slate-500/20 text-slate-400",
  CHECKED_IN: "bg-blue-500/20 text-blue-300",
  COMPLETED:  "bg-emerald-500/20 text-emerald-300",
  MISSED:     "bg-red-500/20 text-red-300",
  SKIPPED:    "bg-amber-500/20 text-amber-300",
};

export const TXN_STATUS_COLORS: Record<TxnStatus, string> = {
  DRAFT:        "bg-slate-500/20 text-slate-400",
  CONFIRMED:    "bg-blue-500/20 text-blue-300",
  INVOICED:     "bg-emerald-500/20 text-emerald-300",
  CANCELLED:    "bg-red-500/20 text-red-400",
  SYNCED:       "bg-indigo-500/20 text-indigo-300",
  PENDING_SYNC: "bg-amber-500/20 text-amber-300",
};

export const PAYMENT_STATUS_COLORS: Record<PaymentStatus, string> = {
  UNPAID:   "bg-red-500/20 text-red-300",
  PARTIAL:  "bg-amber-500/20 text-amber-300",
  PAID:     "bg-emerald-500/20 text-emerald-300",
  REFUNDED: "bg-slate-500/20 text-slate-400",
};

export const RECON_STATUS_COLORS: Record<ReconciliationStatus, string> = {
  OPEN:        "bg-slate-500/20 text-slate-400",
  SUBMITTED:   "bg-blue-500/20 text-blue-300",
  APPROVED:    "bg-emerald-500/20 text-emerald-300",
  DISCREPANCY: "bg-red-500/20 text-red-300",
};

export const SEVERITY_COLORS: Record<string, string> = {
  INFO:     "bg-blue-500/20 text-blue-300",
  WARNING:  "bg-amber-500/20 text-amber-300",
  CRITICAL: "bg-red-500/20 text-red-300",
};

// ── API client ────────────────────────────────────────────────────────────────

export const vanSalesApi = {
  // Vans
  listVans: (params?: { status?: string }) =>
    apiClient.get<Van[]>(`${BASE}/vans`, { params }).then((r) => r.data),
  createVan: (data: Partial<Van>) =>
    apiClient.post<Van>(`${BASE}/vans`, data).then((r) => r.data),
  getVan: (id: string) =>
    apiClient.get<Van>(`${BASE}/vans/${id}`).then((r) => r.data),
  updateVan: (id: string, data: Partial<Van>) =>
    apiClient.patch<Van>(`${BASE}/vans/${id}`, data).then((r) => r.data),

  // Stock
  getStock: (vanId: string) =>
    apiClient.get<VanStock[]>(`${BASE}/vans/${vanId}/stock`).then((r) => r.data),
  loadStock: (vanId: string, data: { lines: Partial<VanStock>[]; notes?: string }) =>
    apiClient.post<VanStock[]>(`${BASE}/vans/${vanId}/stock/load`, data).then((r) => r.data),

  // Visits
  listVisits: (vanId: string, route_date?: string) =>
    apiClient.get<VanVisit[]>(`${BASE}/vans/${vanId}/visits`, { params: { route_date } }).then((r) => r.data),
  createVisit: (vanId: string, data: Partial<VanVisit>) =>
    apiClient.post<VanVisit>(`${BASE}/vans/${vanId}/visits`, data).then((r) => r.data),
  checkIn: (visitId: string, data: { gps_lat?: number; gps_lng?: number }) =>
    apiClient.post<VanVisit>(`${BASE}/visits/${visitId}/check-in`, data).then((r) => r.data),
  checkOut: (visitId: string, notes?: string) =>
    apiClient.post<VanVisit>(`${BASE}/visits/${visitId}/check-out`, {}, { params: { notes } }).then((r) => r.data),
  markMissed: (visitId: string, reason?: string) =>
    apiClient.post<VanVisit>(`${BASE}/visits/${visitId}/missed`, {}, { params: { reason } }).then((r) => r.data),

  // Transactions
  listTransactions: (vanId: string, route_date?: string, params?: { skip?: number; limit?: number }) =>
    apiClient.get<VanSalesTxn[]>(`${BASE}/vans/${vanId}/transactions`, { params: { route_date, ...params } }).then((r) => r.data),
  createTransaction: (vanId: string, data: any) =>
    apiClient.post<VanSalesTxn>(`${BASE}/vans/${vanId}/transactions`, data).then((r) => r.data),
  getTransaction: (txnId: string) =>
    apiClient.get<VanSalesTxn>(`${BASE}/transactions/${txnId}`).then((r) => r.data),

  // Payments
  addPayment: (txnId: string, data: Partial<VanPayment>) =>
    apiClient.post<VanPayment>(`${BASE}/transactions/${txnId}/payments`, data).then((r) => r.data),

  // Reconciliation
  listReconciliations: (vanId: string) =>
    apiClient.get<VanReconciliation[]>(`${BASE}/vans/${vanId}/reconciliations`).then((r) => r.data),
  createReconciliation: (vanId: string, data: { route_date: string; notes?: string }) =>
    apiClient.post<VanReconciliation>(`${BASE}/vans/${vanId}/reconciliations`, data).then((r) => r.data),
  approveReconciliation: (reconId: string) =>
    apiClient.post<VanReconciliation>(`${BASE}/reconciliations/${reconId}/approve`, {}).then((r) => r.data),

  // Sync
  sync: (vanId: string, payload: any) =>
    apiClient.post<SyncResult>(`${BASE}/vans/${vanId}/sync`, payload).then((r) => r.data),

  // Reports
  vanSummary: (vanId: string, days = 30) =>
    apiClient.get<{ txn_count: number; total_sales: number; total_collected: number; days: number }>(
      `${BASE}/reports/van/${vanId}/summary`, { params: { days } }
    ).then((r) => r.data),
  routePerformance: (days = 30) =>
    apiClient.get<{ route_id: string; txn_count: number; total_sales: number }[]>(
      `${BASE}/reports/route-performance`, { params: { days } }
    ).then((r) => r.data),
  driverPerformance: (days = 30) =>
    apiClient.get<{ driver_id: string; txn_count: number; total_sales: number }[]>(
      `${BASE}/reports/driver-performance`, { params: { days } }
    ).then((r) => r.data),

  // AI
  runAI: () => apiClient.post<VSAIRec[]>(`${BASE}/ai/run`, {}).then((r) => r.data),
  aiRecs: (params?: { agent_type?: string; status?: string }) =>
    apiClient.get<VSAIRec[]>(`${BASE}/ai/recommendations`, { params }).then((r) => r.data),
  ackRec: (id: string, status: string) =>
    apiClient.patch<VSAIRec>(`${BASE}/ai/recommendations/${id}`, { status }).then((r) => r.data),
};

File: backend/app/schemas/van_sales.py
from __future__ import annotations
from datetime import date, datetime
from decimal import Decimal
from typing import List, Optional, Dict, Any
from uuid import UUID

from pydantic import BaseModel, Field


# ── Van ───────────────────────────────────────────────────────────────────────

class VanCreate(BaseModel):
    van_code: str
    plate_number: Optional[str] = None
    driver_id: Optional[UUID] = None
    route_id: Optional[UUID] = None
    warehouse_id: Optional[UUID] = None
    capacity_kg: Optional[Decimal] = None
    notes: Optional[str] = None


class VanUpdate(BaseModel):
    plate_number: Optional[str] = None
    driver_id: Optional[UUID] = None
    route_id: Optional[UUID] = None
    warehouse_id: Optional[UUID] = None
    capacity_kg: Optional[Decimal] = None
    status: Optional[str] = None
    notes: Optional[str] = None


class VanOut(BaseModel):
    id: UUID
    van_code: str
    plate_number: Optional[str]
    driver_id: Optional[UUID]
    route_id: Optional[UUID]
    warehouse_id: Optional[UUID]
    capacity_kg: Optional[Decimal]
    status: str
    notes: Optional[str]
    created_at: datetime
    updated_at: datetime
    model_config = {"from_attributes": True}


# ── Van Stock ─────────────────────────────────────────────────────────────────

class VanStockOut(BaseModel):
    id: UUID
    van_id: UUID
    item_id: UUID
    batch_id: Optional[UUID]
    quantity: Decimal
    reserved_qty: Decimal
    unit_cost: Optional[Decimal]
    uom: str
    model_config = {"from_attributes": True}


class StockLoadLine(BaseModel):
    item_id: UUID
    batch_id: Optional[UUID] = None
    quantity: Decimal = Field(gt=0)
    uom: str = "PCS"
    unit_cost: Optional[Decimal] = None


class StockLoadRequest(BaseModel):
    lines: List[StockLoadLine]
    notes: Optional[str] = None


# ── Visit ─────────────────────────────────────────────────────────────────────

class VisitCreate(BaseModel):
    customer_id: UUID
    route_date: date
    sequence_no: Optional[int] = None
    planned_time: Optional[datetime] = None
    offline_id: Optional[str] = None
    notes: Optional[str] = None


class VisitCheckIn(BaseModel):
    gps_lat: Optional[Decimal] = None
    gps_lng: Optional[Decimal] = None


class VisitCheckOut(BaseModel):
    notes: Optional[str] = None


class VisitOut(BaseModel):
    id: UUID
    van_id: UUID
    route_id: Optional[UUID]
    customer_id: UUID
    route_date: date
    sequence_no: Optional[int]
    planned_time: Optional[datetime]
    check_in_time: Optional[datetime]
    check_out_time: Optional[datetime]
    status: str
    gps_lat: Optional[Decimal]
    gps_lng: Optional[Decimal]
    missed_reason: Optional[str]
    offline_id: Optional[str]
    created_at: datetime
    model_config = {"from_attributes": True}


# ── Transaction ───────────────────────────────────────────────────────────────

class TxnLineCreate(BaseModel):
    item_id: UUID
    batch_id: Optional[UUID] = None
    uom: str = "PCS"
    quantity: Decimal = Field(gt=0)
    unit_price: Decimal = Field(ge=0)
    discount_pct: Decimal = Field(ge=0, default=0)
    tax_pct: Decimal = Field(ge=0, default=0)
    return_reason: Optional[str] = None
    notes: Optional[str] = None


class TxnCreate(BaseModel):
    customer_id: UUID
    visit_id: Optional[UUID] = None
    txn_type: str = "SALE"
    txn_datetime: datetime
    lines: List[TxnLineCreate]
    price_list_id: Optional[UUID] = None
    offline_id: Optional[str] = None
    device_id: Optional[str] = None
    notes: Optional[str] = None


class TxnLineOut(BaseModel):
    id: UUID
    txn_id: UUID
    item_id: UUID
    batch_id: Optional[UUID]
    uom: str
    quantity: Decimal
    unit_price: Decimal
    discount_pct: Decimal
    discount_amt: Decimal
    tax_pct: Decimal
    tax_amt: Decimal
    line_total: Decimal
    return_reason: Optional[str]
    model_config = {"from_attributes": True}


class TxnOut(BaseModel):
    id: UUID
    txn_no: str
    van_id: UUID
    route_id: Optional[UUID]
    customer_id: UUID
    visit_id: Optional[UUID]
    driver_id: Optional[UUID]
    txn_type: str
    txn_datetime: datetime
    status: str
    payment_status: str
    subtotal: Decimal
    discount_amount: Decimal
    tax_amount: Decimal
    total_amount: Decimal
    paid_amount: Decimal
    balance_due: Decimal
    currency: str
    linked_so_id: Optional[UUID]
    offline_id: Optional[str]
    synced_at: Optional[datetime]
    notes: Optional[str]
    created_at: datetime
    lines: List[TxnLineOut] = []
    payments: List["PaymentOut"] = []
    model_config = {"from_attributes": True}


# ── Payment ───────────────────────────────────────────────────────────────────

class PaymentCreate(BaseModel):
    method: str
    amount: Decimal = Field(gt=0)
    reference_no: Optional[str] = None
    collected_at: datetime
    notes: Optional[str] = None


class PaymentOut(BaseModel):
    id: UUID
    txn_id: UUID
    method: str
    amount: Decimal
    reference_no: Optional[str]
    collected_at: datetime
    receipt_no: Optional[str]
    notes: Optional[str]
    created_at: datetime
    model_config = {"from_attributes": True}


TxnOut.model_rebuild()


# ── Reconciliation ────────────────────────────────────────────────────────────

class ReconciliationCreate(BaseModel):
    route_date: date
    closing_stock: Optional[Dict[str, float]] = None
    notes: Optional[str] = None


class ReconciliationOut(BaseModel):
    id: UUID
    van_id: UUID
    route_date: date
    status: str
    opening_stock: Optional[Dict]
    closing_stock: Optional[Dict]
    sales_qty: Optional[Dict]
    returns_qty: Optional[Dict]
    discrepancy: Optional[Dict]
    total_sales: Optional[Decimal]
    total_cash: Optional[Decimal]
    total_mpesa: Optional[Decimal]
    total_credit: Optional[Decimal]
    total_returns: Optional[Decimal]
    submitted_by: Optional[UUID]
    approved_by: Optional[UUID]
    approved_at: Optional[datetime]
    notes: Optional[str]
    created_at: datetime
    model_config = {"from_attributes": True}


# ── Offline Sync ──────────────────────────────────────────────────────────────

class SyncPayload(BaseModel):
    visits: List[VisitCreate] = []
    transactions: List[TxnCreate] = []
    payments: List[Dict[str, Any]] = []  # {offline_txn_id, payment}
    device_id: Optional[str] = None


class SyncResult(BaseModel):
    visits_created: int = 0
    txns_created: int = 0
    payments_created: int = 0
    errors: List[str] = []
    duplicates_skipped: int = 0


# ── AI ────────────────────────────────────────────────────────────────────────

class VSAIRecOut(BaseModel):
    id: UUID
    agent_type: str
    status: str
    severity: str
    title: str
    body: str
    data: Optional[dict]
    van_id: Optional[UUID]
    driver_id: Optional[UUID]
    route_id: Optional[UUID]
    created_at: datetime
    model_config = {"from_attributes": True}


class VSAIRecAck(BaseModel):
    status: str
